src/proto/h1/role.rs

```diff
@@ -139,9 +139,10 @@ impl Http1Transaction for Server {
         // but we *never* read any of it until after httparse has assigned
         // values into it. By not zeroing out the stack memory, this saves
         // a good ~5% on pipeline benchmarks.
-        let mut headers_indices = [MaybeUninit::<HeaderIndices>::uninit(); MAX_HEADERS];
+        let mut headers_indices = vec![MaybeUninit::<HeaderIndices>::uninit(); ctx.h1_max_headers];
```

Comment:
This will result in a heap allocation for every request (well, two, the one a couple lines below too). This line is carefully crafted to be fast, which is why it uses uninitialized memory. You can compare the `cargo bench --bench pipeline --features full` of master and this patch to see if there's much change.

I suspect this will require a more involved solution. Perhaps an `enum` that, if the max hasn't been changed, can still use the uninitialized stack array, and if it has, then it uses a `Vec`. (Maybe the `Vec` could even be stored and reused?)

Reply:
https://docs.rs/smallvec/latest/smallvec/index.html I think that's exactly what you want @seanmonstar ?

Reply:
Thinking about it, I'd consider either dropping the configuration, or changing it to contain the pre-allocated number while using smallvec trusting other part of the code (axum for example) to make sure it doesn't exceed header size in kbs and not in count.

Reply:
> https://docs.rs/smallvec/latest/smallvec/index.html I think that's exactly what you want @seanmonstar ?

It seems that there is too little data on the stack and does not meet the requirements 😢

Reply:
> Thinking about it, I'd consider either dropping the configuration, or changing it to contain the pre-allocated number while using smallvec trusting other part of the code to make sure it doesn't exceed header size in kbs and not in count.

I'm also thinking about solutions, maybe const generic has the best performance, but the implementation is ugly.

Using `enum` is also a solution. If you use the default implementation to keep the data on the stack, you can ensure no performance degradation. But if the data is stored on the heap, on the one hand, as @seanmonstar said, it will introduce the overhead of memory application; on the other hand, it will also waste a lot of stack memory, because this huge `enum` uses only a small part on stack.

Reply:
another option - https://github.com/Lokathor/tinyvec

Reply:
> Thinking about it, I'd consider either dropping the configuration, or changing it to contain the pre-allocated number while using smallvec trusting other part of the code to make sure it doesn't exceed header size in kbs and not in count.

both total header size and header count can be useful to limit for different reasons

Reply:
@aviramha Hey I tried `smallvec` and it works great! The code has also been updated!

When using the default value of "max headers", headers are allocated on the stack with no performance penalty. When "max headers" is set, headers will be allocated on the heap and performance will be reduced by approximately 5%.

Perhaps this is the best solution for configurable capacity, cc @seanmonstar

Reply:
> > Thinking about it, I'd consider either dropping the configuration, or changing it to contain the pre-allocated number while using smallvec trusting other part of the code to make sure it doesn't exceed header size in kbs and not in count.
> 
> both total header size and header count can be useful to limit for different reasons

Yeah but I was under the assumption there's other part of the code that validates that, but I might be wrong.